suite/tests/linux/syscall_pwait.c

```diff
@@ -49,40 +49,73 @@
 #include <unistd.h>
 #include <time.h>
 #include <poll.h>
+#include <sched.h>
+#include "thread.h"
+#include "condvar.h"
 
 typedef struct {
     sigset_t *sigmask;
     size_t sizemask;
 } data_t;
 
+typedef struct {
+    pthread_t main_thread;
+    bool delay;
+} args_t;
+
 static struct timespec sleeptime;
+static void *ready_to_listen;
+static args_t a;
 
 static void
 signal_handler(int sig, siginfo_t *siginfo, void *context)
 {
     print("signal received: %d\n", sig);
 }
 
-bool
-kick_off_child_signals()
+THREAD_FUNC_RETURN_TYPE
+kick_off_child_usr1(void *arg)
 {
-    /* waste some time */
-    nanosleep(&sleeptime, NULL);
-
-    int pid = fork();
-    if (pid < 0) {
-        perror("fork error");
-    } else if (pid == 0) {
+    args_t a = *(args_t *)arg;
+    if (a.delay) {
         /* waste some time */
         nanosleep(&sleeptime, NULL);
-        kill(getppid(), SIGUSR2);
+    }
+    pthread_kill(a.main_thread, SIGUSR1);
+    signal_cond_var(ready_to_listen);
+    return THREAD_FUNC_RETURN_ZERO;
+}
+
+THREAD_FUNC_RETURN_TYPE
+kick_off_child_usr2(void *arg)
```

Comment:
Since you're passing a multi-field parameter anyway, I would merge these into one function and pass in the signal.

Reply:
Done.